pkg/controller/common/util.go

```diff
@@ -1619,3 +1620,66 @@ func ProgressFromClaim(ctx context.Context, args *ProgressFromClaimArgs) (string
 
 	return "", nil
 }
+
+// IsSnapshotValidForClone returns true if the passed snapshot is valid for cloning
+func IsSnapshotValidForClone(snapshot *snapshotv1.VolumeSnapshot, log logr.Logger) (bool, error) {
+	if snapshot.Status == nil {
+		log.V(3).Info("Snapshot does not have status populated yet")
+		return false, nil
+	}
+	if snapshot.Status.ReadyToUse == nil || !*snapshot.Status.ReadyToUse {
+		log.V(3).Info("snapshot not ReadyToUse, while we allow this, probably going to be an issue going forward", "namespace", snapshot.Namespace, "name", snapshot.Name)
+	}
+	if snapshot.Status.Error != nil {
+		errMessage := "no details"
+		if msg := snapshot.Status.Error.Message; msg != nil {
+			errMessage = *msg
+		}
+		return false, fmt.Errorf("snapshot in error state with msg: %s", errMessage)
+	}
+	if snapshot.Spec.VolumeSnapshotClassName == nil || *snapshot.Spec.VolumeSnapshotClassName == "" {
+		return false, fmt.Errorf("snapshot %s/%s does not have volume snap class populated, can't clone", snapshot.Name, snapshot.Namespace)
+	}
+	return true, nil
+}
+
+// ValidateSnapshotCloneSize does proper size validation when doing a clone from snapshot operation
+func ValidateSnapshotCloneSize(snapshot *snapshotv1.VolumeSnapshot, pvcSpec *corev1.PersistentVolumeClaimSpec, targetSC *storagev1.StorageClass, log logr.Logger) (bool, error) {
+	restoreSize := snapshot.Status.RestoreSize
+	if restoreSize == nil {
+		return false, fmt.Errorf("snapshot has no RestoreSize")
+	}
+	targetRequest, hasTargetRequest := pvcSpec.Resources.Requests[corev1.ResourceStorage]
+	allowExpansion := targetSC.AllowVolumeExpansion != nil && *targetSC.AllowVolumeExpansion
+	if hasTargetRequest {
+		// otherwise will just use restoreSize
+		if restoreSize.Cmp(targetRequest) < 0 && !allowExpansion {
+			log.V(3).Info("Can't expand restored PVC because SC does not allow expansion, need to fall back to host assisted")
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// ValidateSnapshotCloneProvisioners validates the target PVC storage class against the snapshot class provisioner
+func ValidateSnapshotCloneProvisioners(ctx context.Context, c client.Client, snapshot *snapshotv1.VolumeSnapshot, storageClass *storagev1.StorageClass) (bool, error) {
+	// Do snapshot and storage class validation
+	if storageClass == nil {
+		return false, fmt.Errorf("target storage class not found")
+	}
+	volumeSnapshotContentName := snapshot.Spec.Source.VolumeSnapshotContentName
```

Comment:
`status.boundVolumeSnapshotContentName` is the right place.  The spec field is for manually creating Volumesnapshots.

I hit this when I tried testing this pr manually

Reply:
Thanks! I noticed something wasn't right but couldn't find the fix.